Rolling PR for the clients engine

linacambridge opened the pull request:
The clients engine is used to send and receive commands for sync
engines, like wiping data and resetting Sync metadata. These are
sent in a client record's `commands` list, and acked by removing
the command from the list. The engine syncs twice: once at the
start of the sync, to apply any incoming commands from other clients,
and once at the end, to upload any outgoing commands added as a result
of the sync.

Our implementation is based on Desktop's, but works a little
differently:

* Instead of holding commands in memory, ours pulls outgoing commands
  directly from logins and Places. This follows the approach of
  bookmark restore on Desktop, which writes a flag into Places to send
  a `wipeEngine` command on the next sync.
* Our engine relies on the embedding app to supply it with a `Settings`
  struct containing info about the current client.

Thanks to these two changes, our engine doesn't need to store any state
itself. This means it can be used as a short-lived object that's dropped
after it syncs, like our other syncable stores, and not a long-lived
singleton as on Desktop.

Closes #1380.

### Pull Request checklist ###
<!-- Before submitting the PR, please address each item -->
- [ ] **Quality**: This PR builds and tests run cleanly
  - `cargo test --all` produces no test failures
  - `cargo clippy --all --all-targets --all-features` runs without emitting any warnings
  - `cargo fmt` does not produce any changes to the code
  - `./gradlew ktlint detekt` runs without emitting any warnings
  - `swiftformat --swiftversion 4 megazords components/*/ios && swiftlint` runs without emitting any warnings or producing changes
  - Note: For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes a changelog entry in [CHANGES_UNRELEASED.md](../CHANGES_UNRELEASED.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [ ] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/master/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

linacambridge commented:
@thomcc This is just a start (needs to be wired up to the sync manager, and tests!), but does the design seem sensible to you?

thomcc commented on components/sync_manager/src/manager.rs:
```diff
@@ -38,6 +38,95 @@ impl SyncManager {
     }
 
+    pub fn wipe(&mut self, engine: &str) -> Result<()> {
+        match engine {
```
History?

linacambridge replied:
Added `PlacesApi::wipe_history`.

thomcc commented on components/sync_manager/src/manager.rs:
```diff
@@ -38,6 +38,95 @@ impl SyncManager {
+        {
+            logins.wipe()?;
+        }
+        if let Some(places) = self.places.upgrade() {
```
Should history be here too?

linacambridge replied:
It is now! 😄

thomcc commented on components/sync15/src/clients/ser.rs:
```diff
@@ -0,0 +1,69 @@
+                // Estimate a little more than the direct fraction to maximize packing
+                let cutoff = (list.len() * max_serialized_size - 1) / size + 1;
+                list.truncate(cutoff + 1);
+                // Keep dropping off the last entry until the data fits.
```
Ideally we'd do some sort of binary search here, but in practice this will almost certainly never be hit for clients engine so... whatever.

linacambridge replied:
Yep, agreed. In practice, we'll only expect to hit that case for clients that use legacy Send Tab—only Fennec at this point? All the other commands have predictable sizes, and the only commands we expect to send are wipes if we (eventually?) support bookmark restores.

thomcc commented on components/sync15/src/sync_multiple.rs:
```diff
@@ -192,6 +221,19 @@ impl<'info, 'res, 'pgs, 'mcs> SyncMultipleDriver<'info, 'res, 'pgs, 'mcs> {
+                root_sync_key: &self.root_sync_key,
+            };
+            // Note that a clients engine sync failure is fatal.
+            engine.sync()?;
```
Hmm, I think we'll still want telemetry if we fail here, and it should be easier to add than adding telemetry above (which is before a "sync" so it's unclear where to even put it...).

linacambridge replied:
I'll add this next!

thomcc commented on components/sync_manager/src/manager.rs:
```diff
@@ -249,7 +255,20 @@ impl SyncManager {
+        let settings = Settings {
+            fxa_device_id: params.fxa_device_id,
+            device_name: params.device_name,
+            device_type: if params.device_type == (DeviceType::Mobile as i32) {
```
The list of device types is incomplete now, See https://searchfox.org/mozilla-mobile/source/application-services/components/fxa-client/src/http_client.rs#491-503

linacambridge replied:
I'll go ahead and mirror those here, since we now support fancy new [icons for them](https://searchfox.org/mozilla-central/rev/1d0e84d2ec2843924bc6e5ed6658f93439e12351/browser/base/content/browser-sync.js#882).

linacambridge commented:
(I just rebased, added history wiping, and some tests, but didn't address any of the other feedback yet. I'd also like to add a test for `apply_incoming`, which will involve rejiggering a few things so that we don't have to mock the whole world to make an `Engine`).
